### PR TITLE
Rainbow update (FILE_COUNT,FILE_GET_LIST, FILE_DELETE,FILE_GET_FREE_ID,misc)

### PR DESCRIPTION
**Pull request by BrokeStudio**
FILE_COUNT command added
FILE_GET_LIST command bug fixed
FILE_DELETE command updated
FILE_GET_FREE_ID updated
ASM example code updated
PORT bug fixed in GET_SERVER_SETTINGS

### CODE REVIEW

File: src/boards/rainbow_esp.cpp [hunk 1]
```diff
@@ -365,35 +373,79 @@ void BrokeStudioFirmware::processBufferedMessage() {
...
-				for (uint8 i = 0; i < this->file_exists[path].size(); ++i) {
+				std::vector<uint8> existing_files;
+				this->tx_buffer.push_back(last_byte_read);
+				this->tx_buffer.push_back(2);
+				this->tx_buffer.push_back(static_cast<uint8>(e2n_cmds_t::FILE_COUNT));
+				if (path >= NUM_FILE_PATHS) {
+					this->tx_buffer.push_back(0);
+					break;
+				}
+				assert(this->file_exists[path].size() <= NUM_FILES);
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp]: Strange assert. You are asserting that array size is lower or equal to the constant defining the array size. That should not be a runtime check as it is all compile time values. Actually if it is the intended check, I'd prefere to just not checking it.

Note, it is maybe intended as an update of the check `assert(this->file_exists[path].size() < 254);` which was certainly meant to ensure the following loop using an `uint8` will not overflow. if so, it is also the occasion to remove this overzealous check :)

File: src/boards/rainbow_esp.cpp [hunk 2]
```diff
@@ -365,35 +373,79 @@ void BrokeStudioFirmware::processBufferedMessage() {
...
+				for (uint8 i = 0; i < NUM_FILES; ++i) {
 					if (this->file_exists[path][i]) {
 						existing_files.push_back(i);
 					}
 				}
+				UDBG("%u files found in path %u\n", existing_files.size(), path);
+				this->tx_buffer.push_back(existing_files.size());
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp]: You do not need to fill a vector just to count files. Just increment a counter:
```
uint8 nb_files = 0;
for (bool exists: this->file_exists[path]) {
  if (exist) {
    ++nb_files;
  }
}
this->tx_buffer.push_back(existing_files.size());
```

File: src/boards/rainbow_esp.cpp [hunk 3]
```diff
@@ -365,35 +373,79 @@ void BrokeStudioFirmware::processBufferedMessage() {
 				this->writeFile(this->working_path, this->working_file, this->files[working_path][working_file].size(), this->rx_buffer.begin() + 3, this->rx_buffer.begin() + message_size + 1);
 			}
 			break;
-		case n2e_cmds_t::GET_FILE_LIST:
-			UDBG("RAINBOW BrokeStudioFirmware received message GET_FILE_LIST\n");
+		case n2e_cmds_t::FILE_COUNT:
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp:376]: This command is not in the doc

File: src/boards/rainbow_esp.cpp [hunk 4]
```diff
@@ -365,35 +373,79 @@ void BrokeStudioFirmware::processBufferedMessage() {
...
+				uint8 const path = this->rx_buffer[2];
+				uint8 pageSize = NUM_FILES;
+				uint8 currentPage = 0;
+				if (message_size == 4) {
+					pageSize = this->rx_buffer[3];
+					currentPage = this->rx_buffer[4];
+				}
+				uint8 pageStart = currentPage * pageSize;
+				uint8 pageEnd = currentPage * pageSize + pageSize;
+				uint8 nFiles = 0;
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp]: Coding style: variables are snake_case (camelCase is for functions)

File: src/boards/rainbow_esp.cpp [hunk 5]
```diff
@@ -365,35 +373,79 @@ void BrokeStudioFirmware::processBufferedMessage() {
...
+						nFiles++;
+					}
+					if (nFiles >= pageEnd) break;
+				}
 				this->tx_buffer.push_back(last_byte_read);
 				this->tx_buffer.push_back(existing_files.size() + 2);
 				this->tx_buffer.push_back(static_cast<uint8>(e2n_cmds_t::FILE_LIST));
 				this->tx_buffer.push_back(existing_files.size());
-				for (uint8 i: existing_files) {
+				for (uint8 i : existing_files) {
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp]: Why did you add a space here ?

File: src/boards/rainbow_esp.cpp [hunk 6]
```diff
@@ -454,7 +506,7 @@ uint8 BrokeStudioFirmware::getFreeFileId(uint8 path) const {
 		return NOT_FOUND;
 	}
 	std::array<bool, NUM_FILES> const& existing_files = this->file_exists.at(path);
-	for (size_t i = 0; i < existing_files.size(); ++i) {
+	for (uint8 i = 0; i < static_cast<uint8>(existing_files.size()); ++i) {
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp]: Better keep `size_t` for `i`'s type here. It is used to index an std::array which use `size_t` internally. Forcing `uint8` you just force two more cast by iteration: the one you wrote in the loop's condition and an implicite one when writing `existing_files[i]` as array's operator `[]` takes a `size_t` parameter.

**Review by sgadrat — CHANGES_REQUESTED** [review]

File: src/boards/rainbow_esp.cpp (hunk 3 above)
Comment (BrokeStudio) [reply src/boards/rainbow_esp.cpp:376]: Oops, forgot to commit and push the doc update... Done.

File: src/boards/rainbow_esp.cpp (hunk 5 above)
Comment (BrokeStudio) [reply src/boards/rainbow_esp.cpp]: Absolutely no idea...

File: src/boards/rainbow_esp.cpp (hunk 6 above)
Comment (BrokeStudio) [reply src/boards/rainbow_esp.cpp]: I made this change to make the compiler happy and fix a warning message, but it worked fine before so ok, I'll switch back.

File: src/boards/rainbow_esp.cpp [hunk 7]
```diff
@@ -365,16 +373,52 @@ void BrokeStudioFirmware::processBufferedMessage() {
 				this->writeFile(this->working_path, this->working_file, this->files[working_path][working_file].size(), this->rx_buffer.begin() + 3, this->rx_buffer.begin() + message_size + 1);
 			}
 			break;
-		case n2e_cmds_t::GET_FILE_LIST:
-			UDBG("RAINBOW BrokeStudioFirmware received message GET_FILE_LIST\n");
+		case n2e_cmds_t::FILE_COUNT:
+			UDBG("RAINBOW BrokeStudioFirmware received message FILE_COUNT\n");
 			if (message_size == 2) {
+				uint8 const path = this->rx_buffer[2];
+				std::vector<uint8> existing_files;
```
Comment (sgadrat) [inline src/boards/rainbow_esp.cpp]: unused
